### ISSUE TITLE
Ability to fully control the name of resources

### ISSUE THREAD

**Issue by mikhailshilkov** [pulumi/pulumi#1518]
I've already filed https://github.com/pulumi/pulumi-cloud/issues/504 but this issue is more general.

When I provision Azure resources, they get created with postfixes (random strings appended at the end), e.g. `pulumifunc1sacd7c2fa0`.

I would like to have the ability to define the full name of any resource created by pulumi in full:
- Companies might (and probably should) have policies around naming
- It limits the length available for descriptive names
- Names are then used in things like logging and instrumentation
- Longer bogus names are just harder to read

@joeduffy's initial comment:

> I assume you'd prefer not to have them by default for anything.  It's possible we should make this easier with a global flag or something (`--disableNameMangling`).
> The reason we do it is just so you don't encounter naming collisions when standing up multiple stacks, otherwise people tend to accidentally use names that need to be globally unique, and then only find out when it's "too late" (as in, hundreds of names, and they are trying to stand up a stack in a new region or something).
> But it's very possible we were trying to be too clever here

**Comment by lukehoban** [comment]
Note that you should be able to apply the `name` property to the resources in question to provide exact names (not generated ones).  That should be a workaround in general.

I'll leave this issue open to track the idea of having some sort of global `--disableNameMangling` flag.

**Comment by mikhailshilkov** [comment]
Uh? How do I do that?

**Comment by lukehoban** [comment]
> Uh? How do I do that?

Sorry for the delay.  In general, you can do this with:

```javascript
let x = new Resource("foo", {
  name: "somethingotherthanfoo",
});
```

**Comment by nesl247** [comment]
Any update on this? I don't understand why it defaults to mangling the names, and then having to provide it twice seems strange.

**Comment by lukehoban** [comment]
Some really good additional feedback on this topic at https://github.com/pulumi/pulumi-azure/issues/209#issuecomment-483318254 which should be considered as part of changes here.

**Comment by joeduffy** [comment]
We are planning on doing something here soon. As this has come up several times, I wanted to at least write down the reason Pulumi auto-names resources by appending a suffix, and refer to the documentation. We understand this is confusing and plan to explore options for what we can do here, including opt-in, opt-out, and explicit controls both globally and at the individual resource level.

By auto-appending random suffixes (usually a string of hex), the engine is able to

* Create-before-delete resources during replacement, which minimizes potential downtimes. If names are likely to collide, the engine would need to do the opposite: first, delete the transitive set of resources, then re-provision the new ones. This is far more disruptive than what the engine does now, which resembles a "mini blue/green deployment," at the resource level.

* Reduce the likelihood of naming collisions when standing up multiple stacks. This makes it much easier to stand up new Pulumi environments (e.g., stage, dev, prod, etc), without needing to manually suffix all the names themselves (especially problematic when using components).

These are non-trivial benefits, so whatever we do will need to weigh the tradeoffs involved here. For what it's worth, [AWS CloudFormation does something very similar to this, for similar reasons](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-name.html).

More complete documentation can be found at https://www.pulumi.com/docs/reference/programming-model/#autonaming. Thanks for the patience as we figure out the right path forward!

**Comment by hausdorff** [comment]
@nesl247 can you say more about why is the name mangling on kubernetes irritating?

**Comment by nesl247** [comment]
@hausdorff Since that comment we've actually adopted the auto naming. It was originally irritating as we did everything with expecting pre-determined names.

We've learned a lot about pulumi since then, and why it's beneficial to have the auto naming.

**Comment by joeduffy** [comment]
@lukehoban I'm curious what your latest thinking on this one is.

I'm about to do https://github.com/pulumi/docs/issues/1591 which will at least explain the rationale and behavior here more completely.

For what it's worth, here's where my head is at these days

1. Print a message the first time a stack is stood up and has resources with auto-names in it. Or maybe just do it during `pulumi new`. It could say something like:

    > This stack contains auto-named resources. If you'd like to learn more, or
    > disable this, please visit https://[short url].

2. Add a tooltip ℹ️ indicator in the service UI next to any auto-named resource that says something similar.

3. Add the ability to disable auto-naming for an entire package, e,g. `pulumi config set aws:disableAutoNaming true`. Also add a flag to disable it globally for all packages, e.g. `pulumi config set pulumi:disableAutoNaming true`. Obviously this is a bit of a footgun for non-Kubernetes resources unless we do something like https://github.com/pulumi/pulumi-terraform/issues/383 alongside this.

I realize 1 more so than 2 is a bit unfortunate, because it'll be annoying for end users who want auto-naming, and it's a vast fraction who mind (especially if they're coming from AWS CloudFormation, which does this too). But this has become such a "new user" problem with Pulumi, and is subtle to discover and understand why we do this, that something to raise visibility does seem warranted. Perhaps we could only issue it for your first use of the product?

**Comment by lukehoban** [comment]
Re 1: what I was actually planning on looking at was a better “Welcome to Pulumi” experience on first login that had a little more helpful text generally, but also provided an opportunity for a “tip” style section at the end that could make a note of this.  I am hopeful that we can identify “first login” using existing service APIs such that this isn’t too annoying.

**Comment by lukehoban** [comment]
> unless we do something like pulumi/pulumi-terraform#383 alongside this.  

We are still planning on doing this for 1.0.

**Comment by lukehoban** [comment]
> Add the ability to disable auto-naming for an entire package, e,g. pulumi config set aws:disableAutoNaming true. Also add a flag to disable it globally for all packages, e.g. pulumi config set pulumi:disableAutoNaming true. 

We had not planned to do this part for 1.0, but planned to consider it after we make the other set of changes here post-1.0.

**Comment by joeduffy** [comment]
A member of our community suggested an idea to me yesterday, and I think it might be a winner!

The idea is to expose fine-grained control over auto-naming by way of an interpolated string. For example, our current auto-naming scheme might be represented as `"${name}-${hex(5)}"`. That is to say, the underlying `name`, followed by a `-`, followed by a random hex of 5 digits in length.

This approach has the benefits outlined at https://www.pulumi.com/docs/intro/concepts/programming-model/#autonaming.

Let's say for a moment, however, that you didn't care about any of these. Then you'd simply set the naming interpolation value to `"${name}"`.

Let's say you didn't care about SxS stacks because you're already naming things uniquely, such as using a combination of the project and stack names. But you _do_ care about being able to replace resources without an impactful delete-before-replace teardown that might result in downtime. In that case, we might make it easy to append a sequence number, e.g. `"${name}-v${seqnum}"`.

> An alternative approach to this last point would be to append a blue/green marker of sorts (even/odd) that fluctuates as replacements occur, if you didn't want the full sequence number. The sequence number carries additional benefits, however, because it has semantic meaning and could help you understand what "version" a particular resource is currently at over the lifetime of a stack.

A subtle point here is that the resource name would change during a replacement operation, which I believe would yield a different URN and therefore require engine changes to support.

To control all of these, we could envision a special Pulumi config variable. E.g.

```bash
$ pulumi config set pulumi:nameTemplate "${name}-${hex(5)}"
$ pulumi config set pulumi:nameTemplate "${name}-v${seqnum}"
$ pulumi config set pulumi:nameTemplate "${name}"
...
```

For the common cases, we might add sugar to make it easier to adopt a certain strategy, without needing to explicitly provide the naming template string. For instance, the above might be:

```bash
$ pulumi stack set-naming-strategy autoname
$ pulumi stack set-naming-strategy sequenced
$ pulumi stack set-naming-strategy naked
...
```

Giving people access to the raw interpolated strings seems possibly more powerful than necessary, but I do like that it lets people experiment with strategies that work for them.

**Comment by BrunoRoque06** [comment]
Regarding the naming strategies mentioned, it would be nice to support `snake_case` (with https://www.pulumi.com/docs/intro/concepts/programming-model/#autonaming). When creating Postgres resources for example, and in order to comply to the naming convention, that is the one that should be used.

**Comment by jamesianberry** [comment]
Feedback on my experience with pulumi/examples#611
a) auto-naming is fine most of the time although I went out of my way to ensure a dash separator from the random part that comes along during provisioning (is there a setting for this?)
b) where Azure requires a specific name I allocated both a logical name (without a trailing dash separator) and also e.g. name=AzureFirewallSubnet to ensure the specific physical name which worked fine
c) I had to change my naming convention to ensure maximum name lengths weren't exceeded

**Comment by 0x53A** [comment]
some Azure resources have a very low name length limit, so it would be great if the length of the random part could be controlled - per resource - and reduced from 8 to 6 or 4.

**Comment by Noam-greenberrg** [comment]
Hi 
can any one add a how to disable the autonaming ? 
any option i do is not disable this auto naming ... btw WORST EVER idea :-) 
i test this and not working
config:
  aws:deleteBeforeReplace: "true"
  aws:disableAutoNaming: "true"

Thanks

**Comment by ericrudder** [comment]
Noam, here's some information which might help.

https://www.pulumi.com/blog/infrastructure-as-code-resource-naming/

**Comment by aakrem** [comment]
This is the only main issue still preventing us from choosing Pulumi as a technology to adapt to a Production environment.
The resource names generated by pulumi are ugly, also specifying a resource name and being restrained to edit/replace a resource is definitely not a workaround.

**Comment by joserafaelterra** [comment]
> This is the only main issue still preventing us from choosing Pulumi as a technology to adapt to a Production environment over Terraform.
> The resource names generated by pulumi are ugly, also specifying a resource name and being restrained to edit/replace a resource is definitely not a workaround.


I agree, and not all resources are allowed to customize, **azure_native.containerservice.ManagedCluster** didn't work.

**Comment by mikhailshilkov** [comment]
> not all resources are allowed to customize, azure_native.containerservice.ManagedCluster didn't work

@joserafaelterra  You should be able to set it with the `resourceName` property. If that doesn't work for you, please open an issue with a repro in [azure-native](https://github.com/pulumi/pulumi-azure-native/issues). Thank you!

**Comment by pgavlin** [comment]
I’m wondering if there’s a way to make this work without needing to touch all of our providers.

The larger picture here is that it’s not pulumi per se that is in charge of autonaming. Instead, autonaming is handled on a per-resource basis with each provider. An important aspect of this is that pulumi does not know which properties of which resources are subject to autonaming. As a consequence of this approach, there’s no easy way to change autonaming within pulumi itself--these changes must happen within the providers.

We have made a some recent progress towards making autonames more predictable and presentable: https://github.com/pulumi/pulumi/pull/8631. The title is somewhat misleading, as those changes alone do not make autonames deterministic. We’re still working on consuming these changes from the providers, but once they have been consumed, updated providers will be able to produce friendlier autonames using sequence numbers.

If we want to be more flexible here (ala the changes suggested in https://github.com/pulumi/pulumi/issues/1518#issuecomment-535677365), we’ll probably need to change the provider interface even further, e.g. by adding autoname information to Pulumi Package schemas and using that to handle autonaming within pulumi itself.

If we want to make adjustments here without touching all of the providers, one approach might be to add a pulumi.autoname() API to our SDKs. This API could leave a sentinel value in the resource’s properties that pulumi is able to detect and then replace with an autoname. Parameterizing that API would allow users to control autonaming on a very fine-grain basis at the cost of needing (oddly enough) to explicitly autoname resources.

**Comment by xSAVIKx** [comment]
Just want to add another example that is not directly Pulumi issue, but if the auto-naming could have been configured, I could have avoided this issue as well: https://github.com/GoogleCloudPlatform/buildpacks/issues/172.

TL;DR: some builders require name extension to have e.g. `.zip` if it's a zip archive. And with the auto-naming, even if I specify the resource name with `.zip` the unique hex will be added after the `.zip` breaking the functionality.

**Comment by Frassle** [comment]
> If we want to make adjustments here without touching all of the providers, one approach might be to add a pulumi.autoname() API to our SDKs. This API could leave a sentinel value in the resource’s properties that pulumi is able to detect and then replace with an autoname. Parameterizing that API would allow users to control autonaming on a very fine-grain basis at the cost of needing (oddly enough) to explicitly autoname resources.

If providers could tell us which property was a name as part of the schema we wouldn't even need a special sentinel value here, we could just make it an optional property and if the engine sees it's null it could generate the autoname before passing off to the provider to create.

N.B Even if we changed autonames to work like above I think we still want sequence numbers on Check because providers might want other pseudo-random properties and we don't want the engine to know about every possible one of them.

**Comment by mikhailshilkov** [comment]
> If providers could tell us which property was a name as part of the schema

There are different auto-naming strategies across providers. Azure Native would define whether to add a random suffix based on the resource type (e.g. Cosmos Account is globally unique and thus randomized, but Cosmos Database is unique per server and thus is just named with a straight logical name; Google Native uses composite names with slashes for some resources, etc). If we want to do this in the engine, we'd need to negotiate the autonaming format between the engine, the provider, and (potentially) the user configuration. I guess the schema could have some templated way to describe the auto-name format.

**Comment by Frassle** [comment]
@mikhailshilkov thanks for that didn't realize it was so varying, makes a lot of sense to just leave this in the providers then and deal with that we're going to need to do some updates.

**Comment by lukehoban** [comment]
One option here that is possible today, though not quite as expressive as what is being discussed in other comments above, is to use stack transformations to change the naming strategy for some or all resources.  For example:

```ts
const prefix = `${pulumi.getProject()}-${pulumi.getStack()}-`;

pulumi.runtime.registerStackTransformation(args => {
    if (args.props.name === undefined) {
        return {
            props: {...args.props, name: prefix + args.name },
            opts: args.opts,
        };
    }
    return undefined;
});
```

This will construct the physical name for all resources to be a standard prefix followed by the logical name.  This transformation could also choose to do different things for different resource types (or only apply this for certain resource types).  The one big limitation to this approach is that it cannot gain access to the random value that is part of the current autonaming approach.  Because of that, this approach will lead to delete-before-create replacement semantics (since the physical name is the same, and thus the resource must be deleted before the name can be reused).

**Comment by amuntean-godaddy** [comment]
Is there any chance to have this issue resolved ? I see is stays unresolved for 4 years already.

Auto naming consumes 8 characters, and if the cloud provider has limitation on naming length, it can become a issue to set any meaningful naming.
The ability to control the nr. of characters on random part or at least the ability to access the hash value generated by pulumi in order to use @lukehoban transform method will be helpful.

**Comment by mikhailshilkov** [comment]
@amuntean-godaddy There are a few tradeoffs around this issue, so it keeps slipping through as under-specified. Apologies for delays here.

We are planning to make another attempt to suggest a design proposal for this issue within Q3'22. We will post it here for discussion, and then, hopefully, the actual implementation will follow shortly after.

**Comment by TechWatching** [comment]
@lukehoban [Here](https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/resource-name-rules) is the naming restrictions and rules on Azure we were talking about the other day. Having these rules hard coded in the Azure Native provider would be great to have valid random names automatically.

**Comment by oplancelot** [comment]
> @amuntean-godaddy There are a few tradeoffs around this issue, so it keeps slipping through as under-specified. Apologies for delays here.
> 
> We are planning to make another attempt to suggest a design proposal for this issue within Q3'22. We will post it here for discussion, and then, hopefully, the actual implementation will follow shortly after.

Thank you for your work, but please don't make decisions for the client, give us a global switch, and everyone is happy.

**Comment by Frassle** [comment]
> Thank you for your work, but please don't make decisions for the client

This isn't really about us making decisions for the client, this is about us trying to work out how to even do this from an engineering perspective. We fully understand that this would be a useful feature, we're trying to work out how to do add it without this being overly expensive/complicated to implement.

> give us a global switch, and everyone is happy.

A global switch to turn off autoname suffixes? I don't think that will make everyone happy, there are other requests around naming that are more than just not adding the random suffixes.
Also even that is tricky to add, autonaming is not done in a central location but instead by each provider. So at that point we have to make the decision of if you set "no autonames" how should providers that don't yet understand that option be handled? Given the constraints I think we'd probably just log a warning and then continue to use the provider and get auto-names, but maybe that's not strict enough so now what was a simple "on/off" flag becomes a tri-state of "on/off_warn/off_error", and then we get requests for "I want autonames off and error in general, but I want to allow this one provider through because I know I'm setting explicit names for it". So now we've got a tri-state flag and an exclusion rule list.

That's a lot of engineering effort, and it's not clear it's worth it and I probably haven't even thought of all the other complexities involved here.
We will keep thinking on this problem, but it's not an easy one.

**Comment by oplancelot** [comment]
> > Thank you for your work, but please don't make decisions for the client
> 
> This isn't really about us making decisions for the client, this is about us trying to work out how to even do this from an engineering perspective. We fully understand that this would be a useful feature, we're trying to work out how to do add it without this being overly expensive/complicated to implement.
> 
> > give us a global switch, and everyone is happy.
> 
> A global switch to turn off autoname suffixes? I don't think that will make everyone happy, there are other requests around naming that are more than just not adding the random suffixes. Also even that is tricky to add, autonaming is not done in a central location but instead by each provider. So at that point we have to make the decision of if you set "no autonames" how should providers that don't yet understand that option be handled? Given the constraints I think we'd probably just log a warning and then continue to use the provider and get auto-names, but maybe that's not strict enough so now what was a simple "on/off" flag becomes a tri-state of "on/off_warn/off_error", and then we get requests for "I want autonames off and error in general, but I want to allow this one provider through because I know I'm setting explicit names for it". So now we've got a tri-state flag and an exclusion rule list.
> 
> That's a lot of engineering effort, and it's not clear it's worth it and I probably haven't even thought of all the other complexities involved here. We will keep thinking on this problem, but it's not an easy one.
It's really hard, and you deserve respect.
For example, I want to create  s3 bucket , if the name I want to already exists, I will get a prompt ""Bucket %s exists."", then I will modify it, or modify this in the naming document about the S3 name rules (obviously this one is sloppy), but I don't want to get a random name, because I can't tell the developer to say," please put the image in this bucket that I only know about after building"...it would be weird.
the project requires a lot of collaboration, some information needs to be shared, and most of the time it has to be deterministic.

**Comment by oplancelot** [comment]
```
for _, i
nfra := range infras {

			_, err = helm.NewRelease(ctx, infra, &helm.ReleaseArgs{
				Chart: pulumi.String("../k8s/helm/" + infra),
				// ValueYamlFiles pulumi.AssetOrArchiveArrayOutput `pulumi:"valueYamlFiles"`
				ValueYamlFiles: pulumi.AssetOrArchiveArray{inf_fileAsset1, inf_fileAsset2, inf_fileAsset3, inf_fileAsset4},
				Name:           pulumi.String(infra),
				// Namespace:      pulumi.String("eshop"),
				Values: pulumi.Map{
					"app.name":        pulumi.String(app_name),
					"ingress.hosts":   dns,
					"inf.k8s.dns":     dns,
					"inf.mesh.enable": pulumi.Bool(false),
				},
				// Version: nil,

			})
			if err != nil {
				return err
			}

		}
```
I posted with the above code and it was ok the first time, if I run it again I get the following, actually nothing has changed, I think it's caused by " auto-naming", the google brought me here.
```

ce that already exists. Unable to continue with install: Service "payment-api" in namespace "default" exists and cannot be imported into the current release: invalid ownership metadata; annotation validation error: key "meta.helm.sh/release-name" must equal "payment-api-b91e4e15": current value is "payment-api-3359461c"
```

**Comment by oplancelot** [comment]
> > please put the image in this bucket that I only know about after building"...it would be weird.
> 
> Have you considered using stack outputs to do that? https://www.pulumi.com/learn/building-with-pulumi/stack-outputs/ You can then tell them that the bucket will be deployed as part of a given stack, and that it's id will be output as "bucketID" for example and they can just query the stack output (either via another pulumi stack, via the cli, or direct http request) to see what the ID is.
> 
> It's not deterministic, but it might still solve your information sharing problems.

may be, It's a way, but why can't it be done according to the agreed documentation, some people don't want to write extra code, and more importantly, some development teams don't have permissions to list all S3 buckets.

**Comment by Frassle** [comment]
> may be, It's a way, but why can't it be done according to the agreed documentation, some people don't want to write extra code, and more importantly, some development teams don't have permissions to list all S3 buckets.

For sure it was just a suggestion that _might_ have been applicable in your case. I totally understand that some teams need the deterministic names that they then share via other channels. Scenarios like that are why we understand this issue to be valuable to fix.

> I posted with the above code and it was ok the first time, if I run it again I get the following, actually nothing has changed, I think it's caused by " auto-naming", the google brought me here.

I would suggest raising that issue at https://github.com/pulumi/pulumi-kubernetes/ it might be an issue with helm specifically.

**Comment by sanmai-NL** [comment]
Is [overriding auto-naming as supported](https://www.pulumi.com/docs/intro/concepts/resources/names/#autonaming) not sufficient?

**Comment by sanmai-NL** [comment]
@joeduffy 
> We are planning on doing something here soon. As this has come up several times, I wanted to at least write down the reason Pulumi auto-names resources by appending a suffix, and refer to the documentation. We understand this is confusing and plan to explore options for what we can do here, including opt-in, opt-out, and explicit controls both globally and at the individual resource level.
> 
> By auto-appending random suffixes (usually a string of hex), the engine is able to
> 
> * Create-before-delete resources during replacement, which minimizes potential downtimes. If names are likely to collide, the engine would need to do the opposite: first, delete the transitive set of resources, then re-provision the new ones. This is far more disruptive than what the engine does now, which resembles a "mini blue/green deployment," at the resource level.

That assumes that there's no atomic rename-and-reconfigure operation, or that a separate renaming operation (similar to temporary file naming algorithms) is not a feasible strategy.

> * Reduce the likelihood of naming collisions when standing up multiple stacks. This makes it much easier to stand up new Pulumi environments (e.g., stage, dev, prod, etc), without needing to manually suffix all the names themselves (especially problematic when using components).

If Resources are created in the context of a Resource Component's constructor, then they could auto-name differently. This is a software engineering challenge, not a fundamental problem.

> These are non-trivial benefits, so whatever we do will need to weigh the tradeoffs involved here. For what it's worth, [AWS CloudFormation does something very similar to this, for similar reasons](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-name.html).
> 
> More complete documentation can be found at https://www.pulumi.com/docs/reference/programming-model/#autonaming. Thanks for the patience as we figure out the right path forward!

Okay, so I understand some people don't want to supply the name as resource constructor argument twice. But you can use variables for that, if DRY is the concern.

Personally, the most pressing concern I have is the arbitrary and limiting naming rules that cloud providers impose and how that will interact with any naming algorithm Pulumi devises. For example, Azure limits some resources such as Storage Accounts to 24 characters. 

In terms of declarative style, idempotency etc. a lot can be said for predictable resources names within the provider's namespace. @oplancelot's professional collaboration requirements are of course also entirely realistic, I have them too. Even if the ‘collaboration’ is very advanced, in the form of automated interoperability. Or of the ability for other tools to statically analyze Pulumi source code (as in, resource specs created by users). All of this becomes harder by Pulumi auto-naming opaquely.

I think the whole concept of Stacks is problematic and more trouble than it's worth. Instead of Stacks, the caller could supply a name prefix when standing up, much to the same effect, just no facilities for stack-specific configuration. Because of Stacks, you have had to introduce the concept of ‘fully-qualified Stack names’ (#11387) as well because of collisions.

I think Pulumi currently muddies its functionality between an interactive configuration tool and an infra-as-code-library. It is normal for infra-as-code that things like different environments and their configuration, naming etc. have to be solved by the user (developer). There are many approaches to that depending on the use case. Either static naming, querying some namespace, generating names either human-readable or machine-readable or both, guaranteed unique or just random.

**Comment by sanmai-NL** [comment]
By the way, another interesting problem. Some resources get cloud provider names even as they're not first-class resources in Pulumi. For example, [`FrontendIPConfiguration`](https://www.pulumi.com/registry/packages/azure-native/api-docs/network/loadbalancer/#frontendipconfiguration) in Azure. Pulumi cannot do a delete-after-create in any case here. One may ask, what's the value of delete-after-create if it doesn't cover some important cases anyway?

**Comment by Raniz85** [comment]
To add to the complexity here I would like to adjust the auto-naming strategy.

My current example is an Azure web app with multiple environments with one stack per environment. I would like to use auto-naming so that I get the random suffixes that makes replacement easier, but I would also like to include the stack name in the auto naming so that globally unique resources such as Azure ContainerRegistry doesn't collide between stacks.

At the moment I include the application name and stack name in each resource name to guide the auto naming, like this:

```typescript
const webAppRegistry = new azure.containerregistry.Registry(`MyWebAppRegistry${stack}`, {
    resourceGroupName: resourceGroup.name,
   ...
}
```

But this gets rather tedious and means that all resources have longer names than are necessary. I would rather control it via a callback or format, something like this:

```typescript

const stack = pulumi.getStack()

pulumi.setAutoNamingFormat("MyWebApp{resourceName}" + stack + "{randomSuffix}");

const webAppRegistry = new azure.containerregistry.Registry(`Registry`, {
    resourceGroupName: resourceGroup.name,
   ...
}
```

Where Pulumi would substitute _{resourceName}_ and _{randomSuffix}_ with proper values.

**Comment by rtrindvg** [comment]
I love @Raniz85 suggestion, because I would use a hyphen in the auto naming scheme (to adhere to our naming standards) but right now I need to finish the pulumi object name with a dangling hyphen in the end to achieve this, which is terrible visually.

Also there are some object in some libraries (like azure native app gateway) which needs to refer to it's own name in it's definition, but with the autonaming in place, it becomes impossible. I used an workaround using @pulumi/random but this needs me manually controlling when to rotate the name whenever the resource needs to be replaced.

Having a way to refer to the name after autonaming is in place would help a lot on these cases.

**Comment by gunzy83** [comment]
Nobody else mentioned it here, so I am going to drop my $0.02. 

This might be controversial: I think the Logical and Physical names of resources are two completely different things and should rarely be in any way the same. I will admit that forcing of suffixes to generate physical names of resources based on the logical name does solve some problems, but it also introduces other problems:

* discoverability of resources in the service they are deployed in (due to how generic those names will be)
* extremely long resource names in the Pulumi output if you go the other direction and try to make them identify the resource in a human-readable way

A perfect example of this is an Aurora cluster. I have a project that creates Aurora clusters and RDS instances (there is enough commonality here) and the logical names of the cluster and instances look like this:

* `aurora-cluster`
* `instance-a`
* `instance-b`

If I were to let auto-naming take over from here I would get resource names like this:

* `aurora-cluster-asdfghjk`
* `instance-a-asdfghjk`
* `instance-b-asdfghjk`

Logging into the console and seeing this you would have no idea what DB was what and tools like Teleport would be unusable for our end users (constantly needing to look up the ID to know which DB they are logging into).

Of course there are many compromises like including those identifiers in the logical name (makes reading the Pulumi output just that bit more annoying and less consistent between stacks) but then leaving in superfluous logical items (like `instance`) you end up ugly physical names as well.

I appreciate there are `name` and other options like `namePrefix` (AWS) and `generateName` (k8s, not currently working) that allow better control of naming but it would be extremely valuable to be able to disallow auto-naming based on the logical name to comply with organisation naming standards.

**Comment by rf-0** [comment]
Any update on this?

**Comment by rf-0** [comment]
Any update on this? Even after setting/overriding the names, it seems some functionality is broken.

![image](https://github.com/pulumi/pulumi/assets/43184844/cefb9e1a-f6df-44e7-8c60-f4415690211f)

**Comment by Gonkers** [comment]
I have been using Azure for over 10 years and 4+ years of using Terraform professionally.  I thought I would try Pulumi out personally to see how I would like it. I just stared a little bit ago and already; I found this automatic naming policy to be incredibly irritating. So much so that I don't think I'll bother with Pulumi in the future. I understand putting on the Mickey Mouse gloves for new users because as a new user I want to "fall into the pit of success". But if I want to take off the gloves, please let me manage the automatic naming policy globally.

And the inconsistency is also baffling. The docs say... 
> For cases that require specific names, you can override auto-naming by specifying a physical name. Most resources have a `name` property that you can use to name the resource yourself.

Well that would be better if they all had a `Name` property, but they don't. The custom name property for a `ResourceGroup` is `ResourceGroupName` and the custom name property for a `StorageAccount` is `AccountName`.

**Comment by gunzy83** [comment]
> Well that would be better if they all had a `Name` property, but they don't. The custom name property for a `ResourceGroup` is `ResourceGroupName` and the custom name property for a `StorageAccount` is `AccountName`.

I get your point and the documentation does just say `name` but a heap of this is based on the API of the service behind it (perfect example: S3 buckets use `bucket`/`bucketPrefix` to specify a name and avoid autonaming, this is the same in Terraform btw). I am not sure if that is enough of an annoyance to not consider Pulumi considering Terraform seems to do the same thing.

**Comment by wouter-b** [comment]
Well four years later, the first issue I encountered when starting with pulumi today ;-)

**Comment by ie-zero** [comment]
A suggestion is to have a simple templating system which will allow to specify the allowed characters and the length but everything around it will be treated as literal.

For example, it can be something this which will produce a name `MyWebAppRegistry12345abc`.
```csharp
const webAppRegistry = new azure.containerregistry.Registry(`MyWebAppRegistry<RANDOM_ELEMENT:n5>abc`, {
    resourceGroupName: resourceGroup.name,
   ...
}
```

**Comment by stfnzl** [comment]
@lukehoban nice to see this planned in for Q3 2024! Is the design finalized already?

**Comment by mikhailshilkov** [comment]
Hey folks, I've posted the proposed design up for discussion here: https://github.com/pulumi/pulumi/discussions/17592

We'd appreciate your feedback around it!

**Comment by mikhailshilkov** [comment]
Hey folks! Thank you all for your patience!

I'm excited to say that a preview of this feature has landed in the CLI and a few key providers. To give it a try you would need:

- Pulumi CLI [3.143.0](https://github.com/pulumi/pulumi/releases/tag/v3.143.0) or later
- Pulumi AWS provider [6.66.0](https://github.com/pulumi/pulumi-aws/releases/tag/v6.66.0) (brilliant version number!) or later
- Pulumi Azure Native provider [2.78.0](https://github.com/pulumi/pulumi-azure-native/releases/tag/v2.78.0) or later
- Pulumi Azure Classic provider [6.14.0](https://github.com/pulumi/pulumi-azure/releases/tag/v6.14.0) or later
- Pulumi Google Cloud Platform provider [8.11.0](https://github.com/pulumi/pulumi-gcp/releases/tag/v8.11.0) or later

The feature is currently in preview and hidden behind a configuration flag. You'll have to set `PULUMI_EXPERIMENTAL=1` environment variable to enable it.

Once enabled, you can use the new `pulumi:autonaming` configuration value to set up custom autonaming schemes. A few simple examples:

```yaml
pulumi:autonaming:
  mode: disabled
```

will disable autonaming and require all explicit names.

```yaml
pulumi:autonaming:
  mode: verbatim
```

will copy logical resource names (as in `new MyResource("logicalname", ...)`) to physical cloud resource names as-is, without any random suffixes or other modifications.

```yaml
pulumi:autonaming:
  pattern: ${project}-${stack}-${name}${alphanum(6)}
```

defines a custom pattern that will be populated by the CLI while generating resource names.

You can set autonaming configurations globally (for all providers), for one specific provider, or for one specific resource type:

```yaml
pulumi:autonaming:
    pattern: ${name}-${hex(8)}
    providers:
      aws:
        pattern: ${name}_${alphanum(4)}
        resources:
          "aws:s3/bucket:Bucket": 
            mode: verbatim
```

You can find the full specification in https://github.com/pulumi/pulumi/discussions/17592

The configuration can be set on a project level (Pulumi.yaml), stack level (Pulumi.dev.yaml), or in [ESC](https://www.pulumi.com/docs/esc/).

Our plan is to GA this feature after it rolls out to most providers in January 2025. The docs are coming at the same time.

Until then, please use the `PULUMI_EXPERIMENTAL` flag. We would love to hear your feedback, now is the perfect time to make adjustments based on YOUR experience!

Thank you so much!

**Comment by mikhailshilkov** [comment]
This feature has now officially shipped! You can learn more about it in my blog post: [Introducing Customizable Resource Auto-naming in Pulumi](https://www.pulumi.com/blog/autonaming-configuration/)

Thank you to everyone who provided feedback and helped shape the proposal!

**Comment by Raniz85** [comment]
I've not had a chance to test this yet, but the functionality described in the blog post goes above and beyond my needs when I discovered this ticket.

Well done!
